### PR TITLE
Upgrade to facebook's api v2.6

### PR DESCRIPTION
**Pull request by printercu**
I've received alert from facebook, that current api will be unavailable in august.
> %app% has been making recent API calls to Graph API v2.0, which will reach the end of the 2-year deprecation window on Monday, August 8, 2016. Please migrate all calls to v2.1 or higher in order to avoid potential broken experiences.

Indeed gem seems to be using v1 api, which is alredy deprecated :)

### CODE REVIEW

**Comment by VitorHP** [comment]
👍

**Comment by dsandstrom** [comment]
I received the same email and I have been trying to figure out how to update 2 apps to use the latest Facebook Graph API.

Correct me if I'm wrong, but this gem uses unversioned API calls.  Which, right now uses Facebook Graph's v2.0.  When Facebook removes access to v2.0, requests will use v2.1.  No update to this gem's API version is required.  However, using `json` for the `token_params` might be useful, which I think this PR is accomplishing.

If your app needs to use a certain version of the API (greater than 2.1), you should be able to set the `site` key in the `client_options`.

Regarding the `site` key, it doesn't seem to do anything.  Facebook returns the API version you are accessing in the request's response headers.  However, I can't get it use any version except v2.0.

**Comment by mkdynamic** [comment]
> Correct me if I'm wrong, but this gem uses unversioned API calls. Which, right now uses Facebook Graph's v2.0. When Facebook removes access to v2.0, requests will use v2.1. No update to this gem's API version is required.

That is correct @dsandstrom.

I see the potential confusion this creates, and I think it would be a good change to switch to using versioned APIs. It's more explicit and actually reduces the chance of breakage when Facebook upgrade their APIs since we always know which version the gem is using.

File: lib/omniauth/strategies/facebook.rb [hunk 1]
```diff
@@ -9,18 +9,15 @@ module Strategies
     class Facebook < OmniAuth::Strategies::OAuth2
       class NoAuthorizationCodeError < StandardError; end
 
-      DEFAULT_SCOPE = 'email'
+      DEFAULT_SCOPE = 'email'.freeze
+      API_VERSION = 'v2.6'.freeze
```
Comment (mkdynamic) [inline lib/omniauth/strategies/facebook.rb]: @printercu Can you remove these `.freeze` calls -- inconsistent with other string constants in this gem.

File: test/helper.rb [hunk 2]
```diff
@@ -51,6 +51,10 @@ def strategy
       end
     end
   end
+
+  def http_site
+    strategy.client.site.sub(/^https?/, 'http')
+  end
```
Comment (mkdynamic) [inline test/helper.rb]: @printercu This change seems unrelated/unnecessary to this PR, can you remove these please? (unless I missing why they are needed.)

File: test/strategy_test.rb [hunk 3]
```diff
@@ -75,7 +75,7 @@ class AuthorizeParamsTest < StrategyTestCase
 
 class TokeParamsTest < StrategyTestCase
   test 'has correct parse strategy' do
-    assert_equal :query, strategy.token_params[:parse]
+    assert_equal nil, strategy.token_params[:parse]
   end
 end
```
Comment (mkdynamic) [inline test/strategy_test.rb]: @printercu I think we could remove this entire `TokeParamsTest` now (lol that typo btw. :))

File: lib/omniauth/strategies/facebook.rb [hunk 4]
```diff
@@ -9,18 +9,15 @@ module Strategies
...
 
       option :client_options, {
-        :site => 'https://graph.facebook.com',
+        :site => "https://graph.facebook.com/#{API_VERSION}",
         :authorize_url => "https://www.facebook.com/dialog/oauth",
```
Comment (mkdynamic) [inline lib/omniauth/strategies/facebook.rb]: @printercu I think we should also use the version in this URL, if I am reading the FB docs correctly. Could you make that change, if it seems right?

**Comment by mkdynamic** [comment]
Thanks for this PR @printercu, added a couple comments would be much appreciated if you can address those. Then can merge this and prep a new release (this will be a major version change since is potentially backward incompat.)

File: test/helper.rb (hunk 2 above)
Comment (printercu) [inline test/helper.rb]: @mkdynamic http site is used in images path in several places, take a look for something like:
```
+    assert_equal "#{http_site}/321/picture", strategy.info['image']
```

I thought about adding helper to single test class, but it's used in different ones.

**Comment by printercu** [comment]
@dsandstrom @mkdynamic I've just seen https://developers.facebook.com/docs/apps/changelog, and it says that `/{object}` paths are for v1.0 api. Maybe i'm missing something.

File: lib/omniauth/strategies/facebook.rb (hunk 4 above)
Comment (printercu) [inline lib/omniauth/strategies/facebook.rb]: https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow doesn't use version. See `Invoking the login dialog` section

Comment (printercu) [inline lib/omniauth/strategies/facebook.rb]: Can you please give a link to the docs you read.

File: lib/omniauth/strategies/facebook.rb (hunk 1 above)
Comment (printercu) [inline lib/omniauth/strategies/facebook.rb]: Shouldn't it be better to add `.freeze` to other constants? It would prevent some possible weird cases. Ex. when some other code will use `gsub!` instead of `gsub`, and same of a kind.

Comment (mkdynamic) [inline lib/omniauth/strategies/facebook.rb]: Fair enough, but this change is orthogonal to versioning the API. Goal is to make this diff to minimal. Not a big deal, thanks for taking the effort regardless.

File: lib/omniauth/strategies/facebook.rb (hunk 4 above)
Comment (mkdynamic) [inline lib/omniauth/strategies/facebook.rb]: I'm looking here:

> Versioned paths aren't just true for API endpoints, they're also true for dialogs and social plugins. For example, if you want to generate the Facebook Login dialog for a web app, you can prepend a version number to the endpoint that generates the dialog.
>
> https://developers.facebook.com/docs/apps/versions (under Dialogs)

and in this Gem's README:

https://github.com/mkdynamic/omniauth-facebook#api-version (https://github.com/mkdynamic/omniauth-facebook/commit/9f729037e5c27d102462ebbb205b5da17f9aaf86)

I have not verified it though.

File: test/helper.rb (hunk 2 above)
Comment (mkdynamic) [inline test/helper.rb]: Ditto https://github.com/mkdynamic/omniauth-facebook/pull/245/files#r68485495

**Comment by mkdynamic** [comment]
That seems to contradict the docs here (and also the info in this Gem's README):

> What happens if I don't specify a version for an API?
> 
> We refer to this as an unversioned call. An unversioned call will default to the oldest available version of the API. Consider this hypothetical life cycle of API versions:
> 
> ...
>
> An unversioned call will always point to the oldest version still available at the top of the chart. This is currently v2.0, but after two years it'll be v2.1, then v2.2, etc.
> 
> Because of this, our recommendation is to always specify versions when making calls, where possible.

Clear as mud!

Yet another reason to switch to explicit/versioned endpoints -- will become a moot point then.

File: test/helper.rb (hunk 2 above)
Comment (printercu) [inline test/helper.rb]: Sorry, don't understands. Is it about https://github.com/mkdynamic/omniauth-facebook/pull/245#discussion_r68485495 comment? If I remove this helper this expression will be duplicated 4 times in https://github.com/mkdynamic/omniauth-facebook/pull/245/files#diff-2ae320f27491ec0176a6f06ea4c44cedR102

File: lib/omniauth/strategies/facebook.rb (hunk 4 above)
Comment (printercu) [inline lib/omniauth/strategies/facebook.rb]: Nice catch. I've checked, it works.

**Comment by dsandstrom** [comment]
I agree that not using an api version in the path will use the oldest available version of the api.  This can be proven because this gem doesn't use a explicit version and I can verify it's accessing v2.0.  However, the docs are confusing.

How will developers set the api version they want to use from their app?  Perhaps an option should be added to `provider`.

**Comment by mkdynamic** [comment]
Thanks again for this. Along with a bunch of other updates, I applies this to master -- the next version will use versioned APIs. 4.0.0.rc1 coming soon. Stay tuned.

**Comment by mkdynamic** [comment]
4.0.0.rc1 is out, please check it out and let us know how it goes.

**Comment by dsandstrom** [comment]
Seems to be working.  I figured out how to force a API version when posting to the feed.

I had been using `access_token.post('/me/feed', ...`, figuring the site option would add the hardcoded version number.  However, it's not getting passed through, probably a misunderstanding that I'm making.  What worked for me is using: `access_token.post('/v2.6/me/feed', ...`.

**Comment by mkdynamic** [comment]
@dsandstrom Thanks for this report. Yea, seems like the version prefix should be included in the `access_token`, I'll see about how we can do that.

**Comment by mkdynamic** [comment]
@dsandstrom Can you try without the leading slash? `access_token.post('me/feed', ...)` I think that is what is going on.

**Comment by mkdynamic** [comment]
Yea, that looks like it. Internally the `oauth2` gem uses `faraday` and calls `build_url` with the path:

```ruby
>> require 'faraday'
=> true
>> conn = Faraday::Connection.new
=> #<Faraday::Connection:0x007fa0bb0e9d78 @parallel_manager=nil, @headers={"User-Agent"=>"Faraday v0.9.2"}, @params={}, @options=#<Faraday::RequestOptions (empty)>, @ssl=#<Faraday::SSLOptions (empty)>, @default_parallel_manager=nil, @builder=#<Faraday::RackBuilder:0x007fa0bb0e9b98 @handlers=[Faraday::Request::UrlEncoded, Faraday::Adapter::NetHttp]>, @url_prefix=#<URI::HTTP http:/>, @proxy=nil>
>> conn.url_prefix = "https://sushi.com/api"
=> "https://sushi.com/api"
>> conn.build_url("foo")
=> #<URI::HTTPS https://sushi.com/api/foo>
>> conn.build_url("/foo")
=> #<URI::HTTPS https://sushi.com/foo>
```

**Comment by dsandstrom** [comment]
Thanks Mark, I'll try it tomorrow.  I appreciate you taking time to help me.

**Comment by dsandstrom** [comment]
That's working.  Thanks again.
